src/future/batch_semaphore.rs

```diff
@@ -0,0 +1,606 @@
+
+        state.waiters.remove(index).unwrap();
+        assert!(waiter.is_queued.swap(false, Ordering::SeqCst));
+    }
+
+    /// Acquire the specified number of permits (async API)
+    pub fn acquire(&self, num_permits: usize) -> Acquire<'_> {
+        Acquire::new(self, num_permits)
+    }
+
+    /// Acquire the specified number of permits (blocking API)
+    pub fn acquire_blocking(&self, num_permits: usize) -> Result<(), AcquireError> {
+        crate::future::block_on(self.acquire(num_permits))
+    }
+
+    /// Release `num_permits` back to the Semaphore
+    pub fn release(&self, num_permits: usize) {
+        if num_permits == 0 {
+            return;
+        }
+
+        let mut state = self.state.borrow_mut();
+
+        if ExecutionState::should_stop() {
+            return;
+        }
+
+        // Permits released into the semaphore reflect the releasing thread's
+        // clock; future acquires of those permits are causally dependent on
+        // this event.
+        ExecutionState::with(|s| {
+            let clock = s.increment_clock();
+            state.permits_available.release(num_permits, clock.clone());
+        });
+
+        let me = ExecutionState::me();
+        trace!(task = ?me, avail = ?state.permits_available, waiters = ?state.waiters, "released {} permits for semaphore {:p}", num_permits, &self.state);
+
+        match state.fairness {
+            Fairness::StrictlyFair => {
+                // in a strictly fair mode we will always pick the first waiter
+                // in the queue, as long as there are enough permits available
+                while let Some(front) = state.waiters.front() {
+                    if front.num_permits <= state.permits_available.available() {
+                        let waiter = state.waiters.pop_front().unwrap();
+                        trace!("granted {:?} permits to waiter {:?}", waiter.num_permits, waiter);
+
+                        // The clock we pass into the semaphore is the clock of the
+                        // waiter, corresponding to the point at which the waiter was
+                        // enqueued. The clock we get in return corresponds to the
+                        // join of the clocks of the acquired permits, used to update
+                        // the waiter's clock to causally depend on the release events.
+                        let clock = state
+                            .permits_available
+                            .acquire(waiter.num_permits, waiter.clock.clone())
+                            .unwrap();
+
+                        // Update waiter state as it is no longer in the queue
+                        assert!(waiter.is_queued.swap(false, Ordering::SeqCst));
+                        assert!(!waiter.has_permits.swap(true, Ordering::SeqCst));
+                        ExecutionState::with(|s| {
+                            let task = s.get_mut(waiter.task_id);
+                            if !task.finished() {
```

Comment:
I think this is wrong — we are giving the permits (line 427) to a task which is finished, and which will thus never release the permits. If the task is finished then we can just remove it and give the permits to the next eligible waiter. Also: how does a task become finished while queued for a semaphore?

Reply:
(This is not a change introduced by this PR, but) I agree this seems strange. Changing this to a defensive `assert!(!task.finish())` does not break Shuttle tests at least.